### PR TITLE
[Macros] [SE-0382] Implement macros defined as expansions of other macros

### PR DESCRIPTION
**Pull request by DougGregor**
Implement support for defining a macro in terms of the expansion of another macro, e.g.,

```swift
@freestanding(expression)
macro m1(a: Int) = #m2(a: a, b: 17, c: "Hello")
```

There are a few steps along the way:

* Adopt macro definition checking provided by the `MacroDeclSyntax.checkDefinition` operation (implemented in swift-syntax) in lieu of the checking we currently do on the C++ side
* Only define `$Macros` for compilers that can actually expand macros, and make sure we only run tests that require macro checking for compiler that support expansion.
* Fix an issue with the source ranges for Fix-Its that are bridged from Swift diagnostics over to C++ diagnostics, where we were replacing the leading/trailing trivia of nodes along with the node itself. Only replace the node itself.
* Fix name lookup within the definitions of macros, so we can find the macro parameters.
* Make traversal walk the definitions of macros, rather than ignoring them *cringe*. Improve the AST verifier to better understand macros.

This aspect of macros is tracked by rdar://104044325

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: test/IDE/print_stdlib.swift [hunk 1]
```diff
@@ -2,6 +2,9 @@
 //
 // REQUIRES: nonexecutable_test
 // REQUIRES: swift_swift_parser
+// FIXME: Swift parser is not enabled on Linux CI yet.
+// REQUIRES: OS=macosx
```
Comment (bnbarham) [inline test/IDE/print_stdlib.swift:6]: I assume you ran into whatever I ran into in https://github.com/apple/swift/pull/64089 that caused me not to merge it, though I don't remember what it was 😅

Comment (DougGregor) [reply test/IDE/print_stdlib.swift:6]: YEAH :(

**Pull request merged**